libdrgn/python/program.c

```diff
@@ -477,6 +477,73 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
 	return drgn_object_copy(ret, &((DrgnObject *)obj)->obj);
 }
 
+static struct drgn_error *py_symbol_find_fn(const char *name, uint64_t addr,
+					    enum drgn_find_symbol_flags flags,
+					    void *data, struct drgn_symbol_result_builder *builder)
+{
+	PyGILState_guard();
+
+	_cleanup_pydecref_ PyObject *name_obj = NULL;
+	if (flags & DRGN_FIND_SYM_NAME) {
+		name_obj = PyUnicode_FromString(name);
+		if (!name_obj)
+			return drgn_error_from_python();
+	} else {
+		name_obj = Py_None;
+		Py_INCREF(name_obj);
+	}
+
+	_cleanup_pydecref_ PyObject *address_obj = NULL;
+	if (flags & DRGN_FIND_SYM_ADDR) {
+		address_obj = PyLong_FromUnsignedLong(addr);
+		if (!address_obj)
+			return drgn_error_from_python();
+	} else {
+		address_obj = Py_None;
+		Py_INCREF(address_obj);
+	}
+
+	_cleanup_pydecref_ PyObject *one_obj = PyBool_FromLong(flags & DRGN_FIND_SYM_ONE);
+
+	_cleanup_pydecref_ PyObject *tmp = PyObject_CallFunction(data, "OOO", name_obj,
+								 address_obj, one_obj);
+	if (!tmp)
+		return drgn_error_from_python();
+
+	_cleanup_pydecref_ PyObject *obj =
+		PySequence_Fast(tmp, "Symbol find callback must return a sequence");
+	if (!obj)
+		return drgn_error_from_python();
+
+	size_t len = PySequence_Fast_GET_SIZE(obj);
+	if (len > 1 && (flags & DRGN_FIND_SYM_ONE))  {
+		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+					 "symbol find callback returned multiple elements, but one was requested");
+	}
+
+	for (size_t i = 0; i < len; i++) {
+		PyObject *item = PySequence_Fast_GET_ITEM(obj, i);
+		if (!PyObject_TypeCheck(item, &Symbol_type)) {
+			PyErr_SetString(PyExc_ValueError,
+					"symbol find callback elements must be type Symbol");
+			return drgn_error_from_python();
+		}
+		_cleanup_free_ struct drgn_symbol *sym = malloc(sizeof(*sym));
+		if (!sym)
+			return &drgn_enomem;
+		struct drgn_error *err = drgn_symbol_copy(sym, ((Symbol *)item)->sym);
+		if (err)
+			return err;
+
+		if (!drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym))) {
+			drgn_symbol_destroy(sym);
+			return &drgn_enomem;
+		}
```

Comment:
`no_cleanup_ptr()` sets its argument to `NULL`, so this still leaks. We can do:

```suggestion
		if (!drgn_symbol_result_builder_add(builder, sym))
			return &drgn_enomem;
		sym = NULL; // Owned by the builder now.
```

Reply:
Ah nice catch!